epsagon/events/azure.py

```diff
@@ -0,0 +1,227 @@
+            self.update_response(response)
+
+        if exception is not None:
+            self.set_exception(exception, traceback.format_exc())
+
+    def update_response(self, _response):
+        """
+        Adds response data to event.
+        :param _response: Response from azure
+        :return: None
+        """
+        self.RESPONSE_TO_FUNC.get(self.resource['operation'], empty_func)()
+
+    def set_exception(self, exception, traceback_data, handled=True):
+        """
+        see {Event.set_exception}
+        """
+        super(AzureEvent, self).set_exception(
+            exception,
+            traceback_data,
+            handled=handled
+        )
+
+        # Specific handling for azure errors
+        if isinstance(exception, AzureError):
+            self.event_id = exception.headers.get('x-ms-activity-id')
+            self.resource['metadata']['azure_error'] = True
+            self.resource['metadata']['status_code'] = exception.status_code
+            self.resource['metadata']['error_message'] = exception.message
+            self.resource['metadata']['reason'] = exception.reason
+
+
+class AzureCosmosContainerEvent(AzureEvent):
+    """
+    Represents CosmosDB Container Azure event.
+    """
+
+    RESOURCE_TYPE = 'cosmos_db'
+
+    def __init__(
+            self,
+            wrapped,
+            instance,
+            args,
+            kwargs,
+            start_time,
+            response,
+            exception
+    ):
+        """
+        Initialize.
+        """
+        super(AzureCosmosContainerEvent, self).__init__(
+            wrapped,
+            instance,
+            args,
+            kwargs,
+            start_time,
+            response,
+            exception
+        )
+        self.RESPONSE_TO_FUNC.update({
+            'upsert_item': self.upsert_item_response,
+        })
+
+        self.OPERATION_TO_FUNC.update({
+            'upsert_item': self.upsert_item_request,
+            'delete_item': self.delete_item_request,
+            'query_items': self.query_items_request,
+        })
+        self.kwargs = kwargs
+        self.args = args
+        self.response = response
+
+        self.resource['metadata']['azure.location'] = (
+            instance.client_connection.ReadEndpoint.split(
+                '-'
+            )[1].split('.')[0]
+        )
+        self.resource['metadata']['azure.cosmos.endpoint'] = (
+            instance.client_connection.url_connection
+        )
+        _, database, _, container = instance.container_link.split('/')
+        self.resource['name'] = container
+        self.resource['metadata']['azure.cosmos.database_id'] = database
+        self.resource['metadata']['azure.cosmos.container_id'] = container
+
+        self.OPERATION_TO_FUNC.get(self.resource['operation'], empty_func)()
+
+    def upsert_item_request(self):
+        self.resource['metadata']['azure.cosmos.document_id'] = (
+            self.args[0].get('id')
```

Comment:
Is there always an `id`? if not maybe don't add it to metadata?
(same for delete item operation, and query on query items operation)